packages/teams-js/src/internal/videoUtils.ts

```diff
@@ -78,36 +108,142 @@ function createProcessedStreamGenerator(
   const generator = new MediaStreamTrackGenerator({ kind: 'video' });
   const sink = generator.writable;
 
-  source
-    .pipeThrough(
-      new TransformStream({
-        async transform(originalFrame, controller) {
-          const timestamp = originalFrame.timestamp;
-          if (timestamp !== null) {
-            try {
-              const frameProcessedByApp = await videoFrameHandler({ videoFrame: originalFrame });
-              // the current typescript version(4.6.4) dosn't support webcodecs API fully, we have to do type conversion here.
-              const processedFrame = new VideoFrame(frameProcessedByApp as unknown as CanvasImageSource, {
-                // we need the timestamp to be unchanged from the oirginal frame, so we explicitly set it here.
-                timestamp: timestamp,
-              });
-              controller.enqueue(processedFrame);
-              originalFrame.close();
-              (frameProcessedByApp as VideoFrame).close();
-            } catch (error) {
-              originalFrame.close();
-              notifyError(error);
-            }
-          } else {
-            notifyError('timestamp of the original video frame is null');
-          }
-        },
-      }),
-    )
-    .pipeTo(sink);
+  source.pipeThrough(new TransformStream(transformer)).pipeTo(sink);
   return generator;
 }
 
+class DefaultTransformer {
+  public constructor(private notifyError: (string) => void, private videoFrameHandler: video.VideoFrameHandler) {}
+
+  public transform = async (originalFrame, controller): Promise<void> => {
+    const timestamp = originalFrame.timestamp;
+    if (timestamp !== null) {
+      try {
+        const frameProcessedByApp = await this.videoFrameHandler({ videoFrame: originalFrame });
+        // the current typescript version(4.6.4) dosn't support webcodecs API fully, we have to do type conversion here.
+        const processedFrame = new VideoFrame(frameProcessedByApp as unknown as CanvasImageSource, {
+          // we need the timestamp to be unchanged from the oirginal frame, so we explicitly set it here.
+          timestamp: timestamp,
+        });
+        controller.enqueue(processedFrame);
+        originalFrame.close();
+        (frameProcessedByApp as VideoFrame).close();
+      } catch (error) {
+        originalFrame.close();
+        this.notifyError(error);
+      }
+    } else {
+      this.notifyError('timestamp of the original video frame is null');
+    }
+  };
+}
+
+class TransformerWithMetadata {
```

Comment:
Suggestion: to reduce more repeated code, you may make TransformerWithMetadata and DefaultTransformer both extend from an abstract parent class that has the basic implementation of the `transform` function and an abstract function named `getFrameProcessedByAppFromVideoFrameHandler`. Then TransformerWithMetadata and DefaultTransformer will only need to provide the implementation of `getFrameProcessedByAppFromVideoFrameHandler` without the need to implement the `transform` function

Reply:
seems it won't save more lines, and will be more prune to errors like "videoFrame not closed".